zerver/openapi/javascript_examples.py

```diff
@@ -0,0 +1,168 @@
+
+from typing import Dict, Any, Callable, Set, List
+
+import json
+import os
+import sys
+import subprocess
+
+from functools import wraps
+
+from zerver.openapi.openapi import validate_against_openapi_schema
+
+from zulip import Client
+
+TEST_FUNCTIONS = dict()  # type: Dict[str, Callable[..., None]]
+REGISTERED_TEST_FUNCTIONS = set()  # type: Set[str]
+CALLED_TEST_FUNCTIONS = set()  # type: Set[str]
+
+def openapi_test_function(endpoint: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """This decorator is used to register an openapi test function with
+    its endpoint. Example usage:
+
+    @openapi_test_function("/messages/render:post")
+    def ...
+    """
+    def wrapper(test_func: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(test_func)
+        def _record_calls_wrapper(*args: Any, **kwargs: Any) -> Any:
+            CALLED_TEST_FUNCTIONS.add(test_func.__name__)
+            return test_func(*args, **kwargs)
+
+        REGISTERED_TEST_FUNCTIONS.add(test_func.__name__)
+        TEST_FUNCTIONS[endpoint] = _record_calls_wrapper
+
+        return _record_calls_wrapper
+    return wrapper
+
+def run_js_code(js_code: str) -> List[Dict[str, Any]]:
+    """Here, we run the JavaScript examples and return
+    a list of all the responses from the code example
+    as json."""
+
+    def get_json_data(response_list: List[str]) -> List[Dict[str, Any]]:
+
+        # Converts JavaScript objects to JSON.
+        js_code = """
+var responses = {response_list};
+responses.forEach(
+    (response) => console.log(JSON.stringify(eval("(" + response + ")")))
+);
+""".format(response_list=response_list)
+
+        output = subprocess.check_output(['node'],
+                                         input=js_code,
+                                         universal_newlines=True)
+        json_response_list = [json.loads(response) for response in output.splitlines()]
+
+        return json_response_list
+
+    # Response from 'check_output' here is a string with JavaScript objects
+    # which gets parsed to json responses by 'get_json_data'.
+    responses_str = subprocess.check_output(['node'], input=js_code, universal_newlines=True)
+    json_response_list = get_json_data(responses_str.splitlines())
+
+    return json_response_list
```

Comment:
So, this function is the main thing I think we could do better.  First, I don't understand the layers here (or even why `get_json_data` is defined inside the larger function, rather than as an independent helper), and it'd be nice to have some comments explaining the algorithm here so it's really clear and readable.

Second, I realize this is in a test suite, but in general, we try to avoid using `eval` where possible, for example, especially in this sort of node context running on the developer's system.  

Is there a better way to do this that just has us doing `console.log` and JSON dumping/loading?

Reply:
@timabbott Actually, the js bindings return data in the form of javascript objects instead of json(I don't really understand why.) which cannot be parsed with `json.loads()`. I tried but wasn't able to parse it with python code. `JSON.stringify()` does this job here.

Reply:
@timabbott `run_js_code` runs the example and needs `get_json_data` to parse the output. I defined this inside `run_js_code` as I thought that it wouldn't be needed outside `run_js_code`.
But iI can move it outside if you say.